twenty-first/src/math/polynomial.rs

```diff
@@ -1114,22 +1174,29 @@ where
                 [vec![FF::from(0); n], vec![self.coefficients[0].inverse()]].concat(),
             );
         }
+
         let reverse = self.reverse();
+
         // The next function gives back a polynomial g(X) of degree at most arg,
         // such that f(X) * g(X) = 1 mod X^arg.
         // Without modular reduction, the degree of the product f(X) * g(X) is
         // deg(f) + arg -- even after coefficient reversal. So n = deg(f) + arg
         // and arg = n - deg(f).
         let inverse_reverse = reverse.formal_power_series_inverse_minimal(n - degree);
         let product_reverse = reverse.multiply(&inverse_reverse);
-        product_reverse.reverse()
+        let product = product_reverse.reverse();
+
+        // Coefficient reversal drops trailing zero. Correct for that.
+        let product_degree = product.degree() as usize;
+        product.shift_coefficients(n - product_degree)
     }
 
     /// Reduces f(X) by a structured modulus, which is of the form
     /// X^{m+n} + (something of degree less than m). When the modulus has this
     /// form, polynomial modular reductions can be computed faster than in the
     /// generic case.
     fn reduce_by_structured_modulus(&self, multiple: &Self) -> Self {
+        assert_ne!(multiple.degree(), 0);
```

Comment:
Commit f652255a avoids the `as`-casts using `try_from` and binding the result to a variable. However, I am not convinced about dropping the assert statement because it checks for something different, namely whether the multiple is linear.

Reply:
Good point. Apparently, I misunderstood the `assert!`'s intention.